### PR TITLE
Documentation: Add Benchmark & Profiling Tutorial

### PR DESCRIPTION
**Pull request by sanssecours**
# Description

This pull request contains initial work I did to benchmark the execution time of the various basic YAML plugins. I am pretty sure, that I will extend parts of this PR later. However, since I prefer smaller code updates, I would like to merge this (more or less) in its current state, if the Jenkins build server reports no problems.

### CODE REVIEW

File: doc/tutorials/profiling.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+Since we want to improve the readability of the Callgrind output we choose a build type that includes debug symbols. The two obvious choices for the build type are:
+
+- `RelWithDebInfo` (optimized build with debug symbols), and
+- `Debug` (non-optimized build with debug symbols)
+
+. We use `Debug` here, which should provide the most detailed profiling information.
+
+### Disabling `dlclose` Calls
+
+For this tutorial we decided to profile the [YAy PEG][] plugin. Since Elektra loads plugin code via `dlopen` and Callgrind [does not support the function `dlclose` properly](https://stackoverflow.com/questions/16719395) we remove the `dlclose` calls in the file [`dl.c`](../../src/libs/loader/dl.c) temporarily. At the time of writing one option to do that is deleting
```
Comment (markus2330) [inline doc/tutorials/profiling.md]: What about ENABLE_DEBUG doing this? Or we let ENABLE_DEBUG not tamper with dlclose at all and rename it to ENABLE_ASSERT (see also above).

File: doc/tutorials/benchmarking.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+Usually you want to compare the execution time of the fastest version of a compiled binary. For that purpose it makes sense to change the CMake build type to `Release`, which means that the generated build system will optimize the code and strip debug symbols. You should also disable the logger and debug code. An example CMake command that uses [Ninja](https://ninja-build.org) as build tool could look like this:
+
+```sh
+mkdir build
+cd build
+cmake -GNinja ..                 \
+      -DCMAKE_BUILD_TYPE=Release \
+      -DENABLE_LOGGER=OFF        \
+      -DENABLE_DEBUG=OFF
```
[the remark this answers is not in the document]
Comment (sanssecours) [reply doc/tutorials/benchmarking.md]: I always thought of the variable `ENABLE_DEBUG` as a way to enable additional debug information:

https://github.com/ElektraInitiative/libelektra/blob/0a2520065d628068de6c12cad7a583363ef62558/cmake/ElektraCache.cmake#L263-L268

. For example, [YAMBi](https://www.libelektra.org/plugins/yambi) enables the debug output of the generated parser, if `ENABLE_DEBUG` is set to `ON`:

https://github.com/ElektraInitiative/libelektra/blob/0a2520065d628068de6c12cad7a583363ef62558/src/plugins/yambi/driver.cpp#L148-L150

. If I activate ENABLE_DEBUG the plugin is about 25 times slower.

> Maybe we should rename ENABLE_DEBUG to ENABLE_ASSERT?

If `ENABLE_DEBUG` is **only** a flag that enables assertions, then that would certainly make sense. Currently my plugins also use the macro `DEBUG` (set if `ENABLE_DEBUG` is true) to enable various runtime checks. Other code also uses `DEBUG` to enable additional debug features. For example, the logger changes the default logging level for `syslog`, if the macro is set to true:

https://github.com/ElektraInitiative/libelektra/blob/0a2520065d628068de6c12cad7a583363ef62558/src/include/kdblogger.h#L69-L79
.

File: doc/tutorials/benchmarking.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
+# Benchmarking
+
+## Execution Time
+
+One of the usual questions for a standard benchmark is, how much more or less time two or more different programs take to execute on the same hardware. This tutorial will introduce some tools and techniques that will help you to answer this question. For that purpose we compare the time it takes for certain YAML [storage plugins](storage-plugins.md) to translate YAML data into Elektra’s key set structure. Most of the techniques we describe here should be applicable too, if you want to compare the run-time of other parts of Elektra.
```
[the remark this answers is not in the document]
Comment (sanssecours) [reply doc/tutorials/benchmarking.md]: I added a link to the profiling tutorial in commit c045195. I hope that is a good compromise between 

- me not wanting to write anything 😊, and 
- your desire to create a more complete tutorial

.

File: doc/tutorials/profiling.md (hunk 1 above)
Comment (sanssecours) [reply doc/tutorials/profiling.md]: > What about ENABLE_DEBUG doing this?

While a CMake option to enable this behavior would be nice, as far as I can remember disabling `dlclose` also helps to debug ASAN and Valgrind output, I think `ENABLE_DEBUG` should not introduce additional memory leaks.

**Comment by markus2330** [comment]
Sorry, I produced a merge conflict here, can you please rebase?

**Pull request merged**